st2auth/st2auth/controllers/v1/sso.py

```diff
@@ -0,0 +1,129 @@
+
+from six.moves import http_client
+
+import st2auth.handlers as handlers
+
+from st2auth import sso as st2auth_sso
+from st2common.exceptions import auth as auth_exc
+from st2common import log as logging
+from st2common import router
+
+
+LOG = logging.getLogger(__name__)
+SSO_BACKEND = st2auth_sso.get_sso_backend()
+
+
+class IdentityProviderCallbackController(object):
+
+    def __init__(self):
+        self.st2_auth_handler = handlers.ProxyAuthHandler()
+
+    def post(self, response, **kwargs):
+        try:
+            verified_user = SSO_BACKEND.verify_response(response)
+
+            st2_auth_token_create_request = {'user': verified_user['username'], 'ttl': None}
+
+            st2_auth_token = self.st2_auth_handler.handle_auth(
+                request=st2_auth_token_create_request,
+                remote_addr=verified_user['referer'],
+                remote_user=verified_user['username'],
+                headers={}
+            )
+
+            return process_successful_response(verified_user['referer'], st2_auth_token)
+        except NotImplementedError as e:
+            return process_failure_response(http_client.INTERNAL_SERVER_ERROR, e)
+        except auth_exc.SSOVerificationError as e:
+            return process_failure_response(http_client.UNAUTHORIZED, e)
+        except Exception as e:
+            raise e
+
+
+class SingleSignOnController(object):
+    callback = IdentityProviderCallbackController()
+
+    def get(self, **kwargs):
+        try:
+            referer = kwargs.get('referer')
+            response = router.Response(status=http_client.TEMPORARY_REDIRECT)
+            response.location = SSO_BACKEND.get_request_redirect_url(referer)
+            return response
+        except NotImplementedError as e:
+            return process_failure_response(http_client.INTERNAL_SERVER_ERROR, e)
+        except Exception as e:
+            raise e
+
+
+CALLBACK_SUCCESS_RESPONSE_BODY = """
+<html>
+    <script>
+        function getCookie(name) {
+            var v = document.cookie.match('(^|;) ?' + name + '=([^;]*)(;|$)');
+            return v ? v[2] : null;
+        }
+
+        data = JSON.parse(window.localStorage.getItem('st2Session'));
+        data['token'] = JSON.parse(decodeURIComponent(getCookie('st2-auth-token')));
+        window.localStorage.setItem('st2Session', JSON.stringify(data));
+        window.location.replace("%s");
+    </script>
+    <body/>
+</html>
+"""
+
+
+def process_successful_response(referer, token):
+    token_json = {
+        'id': str(token.id),
+        'user': token.user,
+        'token': token.token,
+        'expiry': str(token.expiry),
+        'service': False,
+        'metadata': {}
+    }
+
+    body = CALLBACK_SUCCESS_RESPONSE_BODY % referer
+    resp = router.Response(body=body)
+    resp.headers['Content-Type'] = 'text/html'
+
+    resp.set_cookie(
+        'st2-auth-token',
+        value=quote(json.dumps(token_json)),
+        expires=datetime.timedelta(seconds=60),
```

Comment:
Just to confirm I understand the flow correctly - st2auth sets a token with a low expire time since this token is directly read by the browser and written into local storage, right?

Reply:
When SSO is successful (response is verified), st2auth will generate a st2 auth token with normal expiration time. In order to pass the st2 auth token from the server side to the client browser, the st2 auth token is written as a cookie with quick expiration time (60 seconds which is set here). The html page sent back to the browser in `CALLBACK_SUCCESS_RESPONSE_BODY` will read the st2 auth token from the cookie and then write it to the local storage.  After 60 seconds, the cookie should be expired and removed by the browser.

Reply:
Yep, thanks, that's what I thought, just wanted to confirm I understand the flow and code correctly.